yellowbrick/utils/kneed.py

```diff
@@ -0,0 +1,208 @@
+
+    def find_knee(self, ):
+        """This function finds and returns the knee value, the normalized knee
+        value, and the x value where the knee is located.
+        :returns: tuple(knee, norm_knee, knee_x)
+        :rtype: (float, float, int)
+        )
+        """
+        if not self.xmx_idx.size:
+            warnings.warn("No local maxima found in the distance curve\n"
+                          "The line is probably not polynomial, try plotting\n"
+                          "the distance curve with plt.plot(knee.xd, knee.yd)\n"
+                          "Also check that you aren't mistakenly setting the curve argument", RuntimeWarning)
+            return None, None, None
+
+        mxmx_iter = np.arange(self.xmx_idx[0], len(self.xsn))
+        xmx_idx_iter = np.append(self.xmx_idx, len(self.xsn))
+
+        knee_, norm_knee_, knee_x = 0.0, 0.0, None
+        for mxmx_i, mxmx in enumerate(xmx_idx_iter):
+            # stopping criteria for exhasuting array
+            if mxmx_i == len(xmx_idx_iter) - 1:
+                break
+            # indices between maxima/minima
+            idxs = (mxmx_iter > xmx_idx_iter[mxmx_i]) * \
+                (mxmx_iter < xmx_idx_iter[mxmx_i + 1])
+            between_local_mx = mxmx_iter[np.where(idxs)]
+
+            for j in between_local_mx:
+                if j in self.xmn_idx:
+                    # reached a minima, x indices are unique
+                    # only need to check if j is a min
+                    if self.yd[j + 1] > self.yd[j]:
+                        self.Tmx[mxmx_i] = 0
+                        knee_x = None  # reset x where yd crossed Tmx
+                    elif self.yd[j + 1] <= self.yd[j]:
+                        warnings.warn("If this is a minima, "
+                                      "how would you ever get here:", RuntimeWarning)
+                if self.yd[j] < self.Tmx[mxmx_i] or self.Tmx[mxmx_i] < 0:
+                    # declare a knee
+                    if not knee_x:
+                        knee_x = j
+                    knee_ = self.x[self.xmx_idx[mxmx_i]]
+                    norm_knee_ = self.xsn[self.xmx_idx[mxmx_i]]
+        return knee_, norm_knee_, knee_x
+
+    def plot_knee_normalized(self, ):
+        """Plot the normalized curve, the distance curve (xd, ysn) and the
+        knee, if it exists.
+        """
+        import matplotlib.pyplot as plt
+
+        plt.figure(figsize=(8, 8))
+        plt.plot(self.xsn, self.ysn)
+        plt.plot(self.xd, self.yd, 'r')
+        plt.xticks(np.arange(min(self.xsn), max(self.xsn) + 0.1, 0.1))
+        plt.yticks(np.arange(min(self.xd), max(self.ysn) + 0.1, 0.1))
+
+        plt.vlines(self.norm_knee, plt.ylim()[0], plt.ylim()[1])
+
+    def plot_knee(self, ):
+        """Plot the curve and the knee, if it exists"""
+        import matplotlib.pyplot as plt
+
+        plt.figure(figsize=(8, 8))
+        plt.plot(self.x, self.y)
+        plt.vlines(self.knee, plt.ylim()[0], plt.ylim()[1])
```

Comment:
These plot functions are probably not necessary, but let's leave them in. Additionally, I'd like to make an issue to refactor this so that these functions take an `ax` argument and plot directly to it, rather than importing `plt` and creating a new figure.